File: cache/src/main/java/com/mm/http/DefaultCallAdapterFactory.java
package com.mm.http;

import android.util.Log;

import androidx.annotation.NonNull;

import java.io.IOException;
import java.lang.annotation.Annotation;
import java.lang.reflect.Method;
import java.lang.reflect.ParameterizedType;
import java.lang.reflect.Type;
import java.util.Iterator;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.concurrent.Executor;

import javax.annotation.Nullable;

import okhttp3.Interceptor;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okio.Timeout;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.SkipCallbackExecutor;

/**
 * Created by : majian
 * Date : 2021/8/24
 */

public class DefaultCallAdapterFactory extends CacheCallAdapter.Factory {
    private final Executor callbackExecutor;

    DefaultCallAdapterFactory(Executor callbackExecutor) {
        this.callbackExecutor = callbackExecutor;
    }

    @Override
    public @Nullable
    CacheCallAdapter<?, ?> get(Type returnType, Annotation[] annotations, RetrofitCache retrofit) {
        if (getRawType(returnType) != Call.class) {
            return null;
        }
        if (!(returnType instanceof ParameterizedType)) {
            throw new IllegalArgumentException(
                    "Call return type must be parameterized as Call<Foo> or Call<? extends Foo>");
        }
        final Type responseType = getParameterUpperBound(0, (ParameterizedType) returnType);

        final Executor executor = Utils.isAnnotationPresent(annotations, SkipCallbackExecutor.class)
                ? null : callbackExecutor;

        return new CacheCallAdapter<Object, Call<?>>() {
            @Override
            public Type responseType() {
                return responseType;
            }

            @Override
            public Call<Object> adapt(Call<Object> call) {
                return executor == null ? call : new ExecutorCallbackCall<>(executor, call);
            }

            @Override
            public okhttp3.Call rawCall(Class<?> service, Method method, Object[] args) {
                Object ser = retrofit.createService(service);
                try {
                    // 删除无效map数据,防止retrofit报错
                    for (Object arg : args) {
                        if (arg instanceof Map) {
                            Iterator<? extends Map.Entry<?, ?>> iterator = ((Map<?, ?>) arg).entrySet().iterator();
                            while (iterator.hasNext()) {
                                Map.Entry<?, ?> entry = iterator.next();
                                if (entry.getKey() == null || entry.getValue() == null) {
                                    iterator.remove();
                                }
                            }
                        }
                    }
                    Object invoke = method.invoke(ser, args);
                    Call<?> call = (Call<?>) invoke;
                    Request request = getRawCallWithInterceptorChain(retrofit, call.request());
                    return retrofit.callFactory().newCall(request);
                } catch (Exception e) {
                    throw new IllegalArgumentException(Log.getStackTraceString(e));
                }
            }
        };
    }

    /**
     * getRawCall from interceptors
     *
     * @param retrofit
     * @param request
     * @return rawRequest
     * @throws IOException
     */
    private Request getRawCallWithInterceptorChain(RetrofitCache retrofit, Request request) throws IOException {
        OkHttpClient client = retrofit.callFactory();
        List<Interceptor> interceptors = client.interceptors();
        int connectTimeoutMillis = client.connectTimeoutMillis();
        int readTimeoutMillis = client.readTimeoutMillis();
        int writeTimeoutMillis = client.writeTimeoutMillis();
        RealRequestInterceptorChain chain = new RealRequestInterceptorChain(client.newCall(request), retrofit.cache, interceptors, 0, request, connectTimeoutMillis, readTimeoutMillis, writeTimeoutMillis);
        return chain.proceed(request).request();
    }


    static final class ExecutorCallbackCall<T> implements Call<T> {
        final Executor callbackExecutor;
        final Call<T> delegate;

        ExecutorCallbackCall(Executor callbackExecutor, Call<T> delegate) {
            this.callbackExecutor = callbackExecutor;
            this.delegate = delegate;
        }

        @Override
        public void enqueue(@NonNull final Callback<T> callback) {
            Objects.requireNonNull(callback, "callback == null");
            delegate.enqueue(new Callback<T>() {
                @Override
                public void onResponse(@NonNull Call<T> call, @NonNull final Response<T> response) {
                    callbackExecutor.execute(() -> {
                        if (delegate.isCanceled()) {
                            callback.onFailure(ExecutorCallbackCall.this, new IOException("Canceled"));
                        } else {
                            callback.onResponse(ExecutorCallbackCall.this, response);
                        }
                    });
                }

                @Override
                public void onFailure(@NonNull Call<T> call, @NonNull final Throwable t) {
                    callbackExecutor.execute(() -> callback.onFailure(ExecutorCallbackCall.this, t));
                }
            });
        }

        @Override
        public boolean isExecuted() {
            return delegate.isExecuted();
        }

        @Override
        public Response<T> execute() throws IOException {
            return delegate.execute();
        }

        @Override
        public void cancel() {
            delegate.cancel();
        }

        @Override
        public boolean isCanceled() {
            return delegate.isCanceled();
        }

        @SuppressWarnings("CloneDoesntCallSuperClone") // Performing deep clone.
        @Override
        public Call<T> clone() {
            return new ExecutorCallbackCall<>(callbackExecutor, delegate.clone());
        }

        @Override
        public Request request() {
            return delegate.request();
        }

        @Override
        public Timeout timeout() {
            return delegate.timeout();
        }
    }
}

File: cache/src/main/java/com/mm/http/DefaultCallAdapterFactory.kt
package com.mm.http

import android.util.Log
import okhttp3.Request
import okio.Timeout
import retrofit2.*
import java.io.IOException
import java.lang.reflect.*
import java.util.*
import java.util.concurrent.Executor
import kotlin.coroutines.Continuation

/**
 * Created by : majian
 * Date : 2021/8/24
 */
class DefaultCallAdapterFactory internal constructor(private val callbackExecutor: Executor) : CacheCallAdapter.Factory() {

    override fun get(returnType: Type, annotations: Array<Annotation>, retrofit: RetrofitCache): CacheCallAdapter<*, *>? {
        if (getRawType(returnType) != Call::class.java) {
            return null
        }
        require(returnType is ParameterizedType) { "Call return type must be parameterized as Call<Foo> or Call<? extends Foo>" }
        val responseType = getParameterUpperBound(0, returnType)
        val executor = if (Utils.isAnnotationPresent(annotations, SkipCallbackExecutor::class.java)) null else callbackExecutor
        return object : CacheCallAdapter<Any, Call<Any>?> {
            override fun responseType(): Type {
                return responseType
            }

            override fun adapt(call: Call<Any>): Call<Any> {
                return if (executor == null) call else ExecutorCallbackCall(executor, call)
            }

            override fun rawCall(service: Class<*>, method: Method, args: Array<Any>, isKotlinSuspendFunction: Boolean): okhttp3.Call {
                val ser = retrofit.createService(service)
                try {
                    // 删除无效map数据,防止retrofit报错
                    args.forEach {
                        if (it is Map<*, *>) {
                            val iterator = it.toMutableMap().entries.iterator()
                            while (iterator.hasNext()) {
                                val next = iterator.next()
                                if (next.key == null || next.value == null) {
                                    iterator.remove()
                                }
                            }
                        }
                    }

                    if (isKotlinSuspendFunction) {
                        //执行retrofit invoke
                        val continuation = args[args.size - 1] as Continuation<*>
                        val clazz = Class.forName("retrofit2.HttpServiceMethod")
                        Proxy.newProxyInstance(clazz.classLoader, clazz.interfaces, object : InvocationHandler {
                            override fun invoke(proxy: Any?, method: Method, args: Array<out Any>): Any {
                                if ("adapt" == method.name) {
                                    args.forEach {
                                        if (it is Call<*>) {
                                            val request = getRawCallWithInterceptorChain(retrofit, it.request())
                                            retrofit.callFactory().newCall(request)
                                        }
                                    }
                                }
                                return method.invoke(proxy, *args)
                            }
                        })
                        //todo 未完成
                        throw IllegalArgumentException("暂不支持该类型的返回数据")
                    } else {
                        val invoke = method.invoke(ser, *args)
                        val call = invoke as Call<*>
                        val request = getRawCallWithInterceptorChain(retrofit, call.request())
                        return retrofit.callFactory().newCall(request)
                    }
                } catch (e: Exception) {
                    throw IllegalArgumentException(Log.getStackTraceString(e))
                }
            }
        }
    }

    /**
     * getRawCall from interceptors
     * @param retrofit
     * @param request
     * @return rawRequest
     * @throws IOException
     */
    @Throws(IOException::class)
    private fun getRawCallWithInterceptorChain(retrofit: RetrofitCache, request: Request): Request {
        val client = retrofit.callFactory()
        val interceptors = client.interceptors
        val connectTimeoutMillis = client.connectTimeoutMillis
        val readTimeoutMillis = client.readTimeoutMillis
        val writeTimeoutMillis = client.writeTimeoutMillis
        val chain = RealRequestInterceptorChain(client.newCall(request), retrofit.cacheHelper, interceptors, 0, request, connectTimeoutMillis, readTimeoutMillis, writeTimeoutMillis)
        return chain.proceed(request).request
    }

    internal class ExecutorCallbackCall<T>(val callbackExecutor: Executor, val delegate: Call<T>) : Call<T> {
        override fun enqueue(callback: Callback<T>) {
            Objects.requireNonNull(callback, "callback == null")
            delegate.enqueue(object : Callback<T> {
                override fun onResponse(call: Call<T>, response: Response<T>) {
                    callbackExecutor.execute {
                        if (delegate.isCanceled) {
                            callback.onFailure(this@ExecutorCallbackCall, IOException("Canceled"))
                        } else {
                            callback.onResponse(this@ExecutorCallbackCall, response)
                        }
                    }
                }

                override fun onFailure(call: Call<T>, t: Throwable) {
                    callbackExecutor.execute { callback.onFailure(this@ExecutorCallbackCall, t) }
                }
            })
        }

        override fun isExecuted(): Boolean {
            return delegate.isExecuted
        }

        @Throws(IOException::class)
        override fun execute(): Response<T> {
            return delegate.execute()
        }

        override fun cancel() {
            delegate.cancel()
        }

        override fun isCanceled(): Boolean {
            return delegate.isCanceled
        }

        override fun clone(): Call<T> {
            return ExecutorCallbackCall(callbackExecutor, delegate.clone())
        }

        override fun request(): Request {
            return delegate.request()
        }

        override fun timeout(): Timeout {
            return delegate.timeout()
        }
    }
}